File: backend/src/common/user-id.ts
import { createHash } from 'node:crypto';

/**
 * Детерминированный маппинг scUserId → number для Qdrant point id.
 * Используется во всех коллекциях user_taste_*, чтобы один юзер имел
 * один и тот же point id во всех векторных пространствах.
 */
export function userIdToQdrantId(userId: string): number {
  const hash = createHash('sha256').update(userId).digest();
  return Number(hash.readBigUInt64BE(0) % BigInt(Number.MAX_SAFE_INTEGER));
}

File: backend/src/common/user_id.rs
use sha2::{Digest, Sha256};

pub const MAX_SAFE_INTEGER: u64 = 9_007_199_254_740_991;

pub fn user_id_to_qdrant_id(user_id: &str) -> u64 {
    let digest = Sha256::digest(user_id.as_bytes());
    let be = u64::from_be_bytes([
        digest[0], digest[1], digest[2], digest[3], digest[4], digest[5], digest[6], digest[7],
    ]);
    be % MAX_SAFE_INTEGER
}
